cmd/guacone/cmd/osv.go

```diff
@@ -84,11 +84,62 @@ var osvCmd = &cobra.Command{
 
 		totalNum := 0
 		var totalDocs []*processor.Document
+		docChan := make(chan *processor.Document)
+		tickInterval := 30 * time.Second
+		ticker := time.NewTicker(tickInterval)
 		gotErr := false
+
+		go func() {
+			for !gotErr {
+				select {
+				case <-ticker.C:
+					if len(totalDocs) > 0 {
+						err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
```

Comment:
The `totalDocs` slice could potentially run into a race condition.

The totalDocs is accessed and modified in the goroutine that handles document ingestion. However, it's also accessed and potentially modified after the wg.Wait() call when the context error is checked and possibly when final ingestion is performed.

If the goroutine is still running and modifying totalDocs when the main goroutine reaches the final ingestion step, a race condition could occur. This is because both goroutines could be trying to access and modify totalDocs at the same time.

Reply:
@naveensrinivasan thanks a lot, it helped me a lot.
I've refactored the whole execution and now `totalDocs` slice is referenced only within the `ingestion` func.

Reply:
Thanks @naveensrinivasan that was a good catch.

Reply:
Thanks. 

- A suggestion is to `break` instead of `flag` for breaking the loop.
- The `gotErr` can also run into a race condition.

```golang
		var gotErr int32
		var wg sync.WaitGroup
		ingestion := func() {
			defer wg.Done()
			var totalDocs []*processor.Document
			const threshold = 1000
		loop:
			for {
				select {
				case <-ticker.C:
					if len(totalDocs) > 0 {
						err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
						if err != nil {
							logger.Errorf("unable to ingest documents: %v", err)
							atomic.StoreInt32(&gotErr, 1)
							break loop
						}
						totalDocs = []*processor.Document{}
					}
					ticker.Reset(tickInterval)
				case d := <-docChan:
					totalNum += 1
					totalDocs = append(totalDocs, d)
					if len(totalDocs) >= threshold {
						err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
						if err != nil {
							atomic.StoreInt32(&gotErr, 1)
							logger.Errorf("unable to ingest documents: %v", err)
							break loop
						}
						totalDocs = []*processor.Document{}
						ticker.Reset(tickInterval)
					}
				case <-ingestionStop:
					break loop
				case <-ctx.Done():
					return
				}
			}
			for len(docChan) > 0 {
				totalNum += 1
				totalDocs = append(totalDocs, <-docChan)
				if len(totalDocs) >= threshold {
					err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
					if err != nil {
						atomic.StoreInt32(&gotErr, 1)
						logger.Errorf("unable to ingest documents: %v", err)
					}
					totalDocs = []*processor.Document{}
				}
			}
			if len(totalDocs) > 0 {
				err = ingestor.MergedIngest(ctx, totalDocs, opts.graphqlEndpoint, csubClient)
				if err != nil {
					atomic.StoreInt32(&gotErr, 1)
					logger.Errorf("unable to ingest documents: %v", err)
				}
			}
		}
```

Reply:
@naveensrinivasan thanks a lot :pray:

Reply:
@mrizzi  Appreciate taking the feedback 👍

Reply:
seems like the flag is a better option for breaking to loop instead of a goto statement

Reply:
:sweat_smile: 
So, I have no strong opinion on this guys, but I'm concerned the certifier OSV polling is broken.

The only rationale to decide how to implement this that I can see is the consistency with the rest of guac. It looks to me that flag approach is widely used.

@naveensrinivasan @pxp928 WDYT?

Reply:
I am good :)

Reply:
Yeah, let's stick with the flag. Thanks @mrizzi